src/services/executionEnvironment.ts

```diff
@@ -0,0 +1,309 @@
+import { ImagePuller } from '../utils/imagePuller';
+import { asyncExec } from '../utils/misc';
+import { WorkspaceFolderContext } from './workspaceManager';
+import { IContainerEngine } from '../interfaces/extensionSettings';
+
+export class ExecutionEnvironment {
+  private connection: Connection;
+  private context: WorkspaceFolderContext;
+  private _container_engine: IContainerEngine;
+  private _container_image: string;
+  private _container_image_id: string;
+
+  constructor(connection: Connection, context: WorkspaceFolderContext) {
+    this.connection = connection;
+    this.context = context;
+  }
+
+  public async initialize(): Promise<void> {
+    try {
+      const settings = await this.context.documentSettings.get(
+        this.context.workspaceFolder.uri
+      );
+      this._container_image = settings.executionEnvironment.image;
+      this._container_engine = settings.executionEnvironment.containerEngine;
+      if (this._container_engine === 'auto') {
+        for (const ce of ['podman', 'docker']) {
+          try {
+            child_process.execSync(`which ${ce}`, {
+              encoding: 'utf-8',
+            });
+          } catch (error) {
+            this.connection.console.info(`Container engine '${ce}' not found`);
+            continue;
+          }
+          this._container_engine = <IContainerEngine>ce;
+          this.connection.console.log(`Container engine set to: '${ce}'`);
+          break;
+        }
+      } else {
+        try {
+          child_process.execSync(`which ${this._container_engine}`, {
+            encoding: 'utf-8',
+          });
+        } catch (error) {
+          this.connection.window.showErrorMessage(
+            `Container engine '${this._container_engine}' not found. Failed with error '${error}'`
+          );
+          return;
+        }
+      }
+      if (!['podman', 'docker'].includes(this._container_engine)) {
+        this.connection.window.showInformationMessage(
+          'No valid container engine found.'
+        );
+        return;
+      }
+      const imagePuller = new ImagePuller(
+        this.connection,
+        this.context,
+        this._container_engine,
+        this._container_image,
+        settings.executionEnvironment.pullPolicy
+      );
+      const setupDone = await imagePuller.setupImage();
+      if (!setupDone) {
+        this.connection.window.showErrorMessage(
+          `Execution environment image '${this._container_image}' setup failed.
+           For more details check output console logs for ansible-language-server`
+        );
+        return;
+      }
+      this.fetchPluginDocs();
+    } catch (error) {
+      if (error instanceof Error) {
+        this.connection.window.showErrorMessage(error.message);
+      } else {
+        this.connection.console.error(
+          `Exception in ExecutionEnvironment service: ${JSON.stringify(error)}`
+        );
+      }
+    }
+  }
+
+  async fetchPluginDocs(): Promise<void> {
```

Comment:
This is likely to take a longer time (at least the first time). A progress tracker should be employed here, so that user is aware that something is happening behind the scenes.

Reply:
fixed